Antistasi Zeus Units, Groups, Vehicles, Modules

jwoodruff40 opened the pull request:
## What type of PR is this?
- [ ] Bug
- [ ] Change
- [x] Feature
- [x] Miscellaneous

<details><summary><i><b>
Sub-categories:
</b></i></summary>

- [ ] Template
- [ ] Map
- [x] Config
- [ ] Function
- [ ] Localization

</details>

## What have you changed, and why?

**Information:**

> Adds Antistasi units, vehicles, groups to Zeus
>
> Currently supports:
> - Rebels units, vehicles, and groups
> - Occupants units, vehicles, and groups
> - Invaders units, vehicles, and groups
> - Rivals units and vehicles
> - Civilian units and vehicles
> - Modules for Occupant and Invader supports (from fn_initSupports.sqf)
> - Module to add rebel group to high command
>
> Todo:
> - rival groups? not sure what the composition should be...
> - update faction icons? eh does it really matter?
> - ~~Figure out a way to choose side for supports (currently hardcoded in modules, supporting Occupants only unless duplicated for invaders)~~
>
> Notes:
>
> ~~For the life of me, I can't figure out why vehicles won't show up in units in zeus, but units and groups work just fine. The groups use the same vehicle classes that aren't showing in units...~~
>
> ~~This PR is *NOT* ready. I am only submitting as a draft because I happened across https://github.com/official-antistasi-community/A3-Antistasi/pull/3686 today while looking for something else, and I'm not interested in getting accused of 'stealing' code later; that said, ante did some things that probably are better architected than this (zeus units in a new addon, having unit postinit call a dedicated init function to save code duplication in all the unit init event handlers, etc. So I'll probably copy those *design* decisions at some point.~~

**How can your changes be tested, or reproduced?**

> Use zeus while in an Antistasi game

**Does this PR include changes not authored by you?**

- [x] No
- [ ] Yes (See below)

<details><summary><i><b>
Further info:
</b></i></summary><br>

- [ ] I confirm that I, and by extension this repository, can legally use these third-party changes. (Provide links or author attribution.)

> ...

</details>

## Please verify the following (If possible).

`*` - Mandatory

- [x] These changes are my own, or I have written permission to use them. `*`
- [x] These changes are ready for review, or will be marked as a draft. `*`
- [x] I have loaded the mission in LAN host.
- [x] I have loaded the mission on a dedicated server.

Is further work needed?

- [x] Needs further testing.
- [x] Needs further changes.
- [ ] Needs to be converted to a draft.

## Please specify which Issue this PR Resolves (If Applicable).
This PR closes #XYZ.

<hr><details><summary><i><b>
Notes:
</b></i></summary><br>

> ...

</details>

jwoodruff40 committed:
A3U Editor Factions basic structure test

jwoodruff40 committed:
Add all rebel and occupant soldiers

UnseenKill commented on A3A/addons/ultimate/patches/cfgVehicles.hpp:
```diff
@@ -261,6 +261,928 @@ class CfgVehicles {
+            class ADDON : ADDON {
+                init = "params ['_unit']; _unit setVariable ['unitType', 'loadouts_reb_militia_Rifleman'];";
+            };
+        };
```
Replace all `EventHandlers` with this one line:

```suggestion
        GVAR(unitType) = QUOTE(loadouts_reb_militia_Rifleman);
```

UnseenKill commented on A3A/addons/ultimate/patches/cfgVehicles.hpp:
```diff
@@ -261,6 +261,928 @@ class CfgVehicles {
+            class ADDON {
+                PostInit = "params ['_unit']; [_unit, false] call A3A_fnc_fiaInit;";
+            };
+        };
```
The base class can do all the repetitive work you're doing in the child classes:

```suggestion
        class EventHandlers {
            class ADDON {
                PostInit = QUOTE([ARR_2(_this#0,false)] call A3A_fnc_fiaInit);
                init = QUOTE((_this#0) setVariable[ARR_2(QUOTE(QUOTE(unitType)),getText(configOf(_this#0) >> QQGVAR(unitType)))]);
            };
        };
```

jwoodruff40 replied:
Whaaaaaaaa.

I'ma need some schooling on this GVAR stuff. I did not like the init / postinit EHs either, but couldn't figure out how to set an arbitrary var on the units in config.

Does GVAR do the same thing in config as `setVariable` in sqf? Or similar idea anyway? That's definitely a lot cleaner if so. I mainly just didn't want to write a whole separate init func for the units and reuse the existing fiaInit / NATOinit which rely on variables.

UnseenKill replied:
> Does GVAR do the same thing in config as setVariable in sqf?

No, not at all. The macro is used to prefix (global) variables or things that could conflict with other mods. `GVAR(foo)` is basically `A3A_foo` in this case.

In [this suggestion](https://github.com/Antistasi-Ultimate-Community/A3-Antistasi-Ultimate/pull/681/files#r2513712053), `GVAR(unitType)` (i.e. `A3A_unitType`) is a property for the child class. Since the child class uses its parent class' `init` code (from this suggestion), it can read the property.

jwoodruff40 replied:
Ok, I think I get it.

Will commit the changes for the rebels, and then play with doing something similar for the occupants / enemy factions. Got an idea of how to clean those up in a similar vein.

UnseenKill committed:
Merge branch 'unstable' into edenzeus-factions

jwoodruff40 committed:
unseen is smarter than me

Co-authored-by: UnseenKill <post@oliver-schieche.de>
Signed-off-by: John Woodruff <22091381+jwoodruff40@users.noreply.github.com>

UnseenKill commented on A3A/addons/ultimate/patches/cfgVehicles.hpp:
```diff
@@ -261,6 +261,925 @@ class CfgVehicles {
+            class ADDON : ADDON {
+                init = "params ['_unit']; _unit setVariable ['unitType', 'loadouts_reb_militia_SquadLeader'];";
+            };
+        };
```
```suggestion
        GVAR(unitType) = QUOTE(loadouts_reb_militia_SquadLeader);
```

UnseenKill commented on A3A/addons/ultimate/patches/cfgVehicles.hpp:
```diff
@@ -261,6 +261,925 @@ class CfgVehicles {
+            class ADDON : ADDON {
+                init = "params ['_unit']; _unit setVariable ['unitType', 'loadouts_reb_militia_staticCrew'];";
+            };
+        };
```
```suggestion
        GVAR(unitType) = QUOTE(loadouts_reb_militia_staticCrew);
```

UnseenKill commented on A3A/addons/ultimate/patches/cfgVehicles.hpp:
```diff
@@ -261,6 +261,925 @@ class CfgVehicles {
+            class ADDON : ADDON {
+                init = "params ['_unit']; _unit setVariable ['unitType', 'loadouts_reb_militia_Medic'];";
+            };
+        };
```
```suggestion
        GVAR(unitType) = QUOTE(loadouts_reb_militia_Medic);
```

UnseenKill replied:
etc. This should be the only line in any class.

jwoodruff40 committed:
Merge remote-tracking branch 'upstream/unstable' into edenzeus-factions

jwoodruff40 committed:
rebase (not an actual rebase) on 11.8.6 unstable

jwoodruff40 committed:
move existing zeus work to dedicated addon

jwoodruff40 committed:
Add dedicated init function for all units;

cleanup gross init / postInit event handlers;

add all invader and occupant units

jwoodruff40 commented:
Moved to a new addon (need to add `A3A/addons/zeus` to clientDirs when building).

Got rid of the gross init / postInit handlers in all the units with a dedicated function that handles init for all units. Still using GVARs for unit faction, prefix, and type to accomodate.

All factions except civilians added to individual units. Still need to build groups (or more likely, another init handler for spawning the group using existing functions...)

jwoodruff40 committed:
zeus stringtable

jwoodruff40 committed:
Add rebel vehicles and groups;

restructure CfgVehicles

jwoodruff40 committed:
add occupant and invader vehicles;

misc fixes and updates

jwoodruff40 committed:
misc vehicle fixes across factions

jwoodruff40 committed:
occupant and invader infantry groups

jwoodruff40 committed:
change stringtable prefix from A3AU to A3U

jwoodruff40 committed:
Update Stringtable.xml

jwoodruff40 committed:
add rival vehicles;

fix classnames for occupants groups

jwoodruff40 committed:
Update A3U_Rivals.hpp

jwoodruff40 committed:
add civilians;

update inheritance for groups

jwoodruff40 committed:
Merge branch 'edenzeus-factions' of https://github.com/jwoodruff40/A3-Antistasi-Ultimate into edenzeus-factions

jwoodruff40 committed:
ugh fixes

jwoodruff40 committed:
Update A3U_Rivals.hpp

jwoodruff40 committed:
Rework vehicle crewing using existing A3A_fnc_createVehicleCrew;

Add failure message when vehicles of given type don't exist in the template

jwoodruff40 committed:
Merge branch 'unstable' into edenzeus-factions

jwoodruff40 committed:
Fix APC display names

jwoodruff40 committed:
Fix multiple vehicle spawning on dedicated

jwoodruff40 committed:
Fix invader groups display name

jwoodruff40 committed:
'Place vehicles with crew' checkbox implementation

jwoodruff40 committed:
macros

jwoodruff40 committed:
formatting

jwoodruff40 committed:
Antistasi Zeus modules groundwork

jwoodruff40 committed:
Initial working support module + initModule handler

jwoodruff40 committed:
add most supports; occupants only

jwoodruff40 committed:
Merge remote-tracking branch 'upstream/unstable' into edenzeus-factions

jwoodruff40 committed:
working module attachment

jwoodruff40 committed:
Merge remote-tracking branch 'upstream/unstable' into edenzeus-factions

jwoodruff40 committed:
Rework zeus error messages;

add error message if not in an antistasi ultimate (or plus) mission

jwoodruff40 committed:
invader supports and general rework of initModule

UnseenKill commented on A3A/addons/zeus/script_component.hpp:
```diff
@@ -0,0 +1,4 @@
+#define COMPONENT zeus
+#define PREFIX A3U
+#include "\x\A3A\addons\core\Includes\script_mod.hpp"
+#define PATCHNAME(x) ADDON_ZEUS##_##x
```
`PATCHNAME` macro is unused and will become unnecessary with #770.

```suggestion
#include "\x\A3A\addons\core\Includes\script_mod.hpp"

```

jwoodruff40 committed:
Update A3A/addons/zeus/script_component.hpp

Co-authored-by: UnseenKill <post@oliver-schieche.de>
Signed-off-by: John Woodruff <22091381+jwoodruff40@users.noreply.github.com>

UnseenKill commented on A3A/addons/zeus/script_component.hpp:
```diff
@@ -0,0 +1,3 @@
+#define COMPONENT zeus
+#define PREFIX A3U
```
```suggestion
#define PREFIX A3A
```

Doesn't load functions otherwise.

jwoodruff40 replied:
I'm not understanding the issue here. What functions are not loaded? The functions are deliberately `A3U_zeus_fnc_xxx`. Everything is working as intended afaict.

UnseenKill replied:
Your `$PBOPREFIX$` says, it's supposed to be `A3A`.

By luck, your `PREFIX` gets overwritten with `MODFOLDER`, a previously unused macro, that'll die with #770, though. That makes function name to path solution work for now.

So functions are called `A3U_fnc_*` but are _read_ from `\x\A3A\...`.

So... Since we're not using `A3U` prefix (as a macro) anywhere else, better to change it, before it blows up with #770.

jwoodruff40 replied:
Ahhhh I see. I blame that quiet guy that created the `ultimate` addon because I copied most of the boilerplate from there lol. Will update later.

Silence-112 replied:
![cat](https://github.com/user-attachments/assets/6effecec-d65c-4c19-8b2e-b03e48605002)

wersal454 replied:
> ![cat](https://private-user-images.githubusercontent.com/78276788/547177545-6effecec-d65c-4c19-8b2e-b03e48605002.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzA2NTMyMzMsIm5iZiI6MTc3MDY1MjkzMywicGF0aCI6Ii83ODI3Njc4OC81NDcxNzc1NDUtNmVmZmVjZWMtZDY1Yy00YzE5LThiMmUtYjAzZTQ4NjA1MDAyLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjA5VDE2MDIxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdiZGVlM2VhNWQ4YjM4M2NmMzMzMzQzMzQyMjE5NGM3MTk5ZGNmN2MzMTYxNWYwMzQzODdhNDU3YzNlM2VmZGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XQnRKyNNvt366KKJUlCMaZgEsjWjD6_HdK__AAWuMyk)

r*

jwoodruff40 committed:
Rival vehicle crew fix

jwoodruff40 committed:
Add to high command module

jwoodruff40 committed:
prefix change

jwoodruff40 committed:
misc civ unit fixes

jwoodruff40 committed:
Remove AT vehicles from Occ/Inv
